src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -176,124 +176,194 @@ class CompilationState( statements : DeepDiveLog.Program, config : DeepDiveLog.C
     if( ground_relations contains relName ) !ground_relations(relName) else true
   }
 
-  // Resolve a column name with alias
-  def resolveColumn(s: String, q : ConjunctiveQuery, alias: AliasStyle) : Option[String] = {
-    val qs = new QuerySchema(q)
-    val index = qs.getBodyIndex(s)
-    val name  = resolveName(qs.getVar(s))
-    val relation = q.bodies(0)(index).name
-    alias match {
-      case OriginalOnly => Some(s"R${index}.${name}")
-      case AliasOnly => Some(s"${relation}.R${index}.${name}")
-      case OriginalAndAlias => Some(s"""R${index}.${name} AS "${relation}.R${index}.${name}" """)
+  // compile a variable name to its canonical relation and column in R{index}.{name} style
+  def compileVariable(v: String, cq: ConjunctiveQuery) = {
+    val qs = new QuerySchema(cq)
+    val index = qs.getBodyIndex(v)
+    val name  = resolveName(qs.getVar(v))
+    s"R${index}.${name}"
+  }
+
+  // compile alias, only used for head
+  def compileAlias(e: Expr, cq: ConjunctiveQuery, alias: AliasStyle) = {
+    e match {
+      case VarExpr(v) => {
+        val qs = new QuerySchema(cq)
+        val index = qs.getBodyIndex(v)
+        val variable = qs.getVar(v)
+        val name  = resolveName(variable)
+        val relation = variable.relName
+        alias match {
+          case NoAlias => ""
+          case HeadAlias => s""" AS "${relation}.R${index}.${name}" """
+          case WeightAlias => s"${relation}.R${index}.${name}"
+        }
+      }
+      case _ => s" AS column_${cq.head.terms indexOf e}"
     }
   }
 
   // resolve an expression
-  def compileExpr(e: Expr, cq: ConjunctiveQuery, alias: AliasStyle, index : Int, isHead: Boolean) : String = {
+  def compileExpr(e: Expr, cq: ConjunctiveQuery, generateAlias: Expr => String = e => "") = {
```

Comment:
I think taking a substitution function of type `String => String` (that corresponds to `compileVariable(_, cq)`) as the only extra parameter makes more sense here because it'll allows dropping `cq`, which is never used elsewhere.

Since `generateAlias` isn't doing anything inside `compileExprInner`, we should also take that completely out to the callers.  If you look at how `generateAlias` is being instantiated, it's all some form of `compileAlias`.  If that's the case, `compileAlias` can actually call `compileExpr`, appending the `" AS ..."`, and `compileExpr` can be as simple as `(Expr, String => String) => String`.

Could you revise the refactoring if these make sense?